### DISCUSSION TITLE
Backward compatible drop column flow while running alembic migrations concurrently to application code

### DISCUSSION THREAD

**Discussion by ilanKeshet** [sqlalchemy/sqlalchemy#8148] [Usage Questions]
Hi All, 

First of all, Thank you! 
We ❤️ SQLAlchemy and use it every day.

We are running a HA application in a microservice architecture using SQLAlchemy to connect to a database, and Alembic to run migrations. 
Currently using `sqlalchemy==1.3.24` and `alembic==1.7.4` (planning to upgrade to 1.4 and 1.8 in the near future) to access a postgresql database.
The thing is that we can't stop the application in order to run migrations and so all of our schema-changes are backward compatible. This is needed for migrations for a _new_ version to be executed in a parallel to the _old_ application code. 
Another this is worth noting is that we tests validating that the Models (Entities?, ORM declarative classes) as they exist presently and database schema as generated using alembic replaying the entire history match. 

**The flow we currently have is as follows:**
1. Remove from code all usage of a model column
2. Delete th column from the model (the ORM declarative class) 
3. Pass Alembic an `include_object` ignoring changes to the column (so that it won't generate a drop column schema migration for it just yet) 
4. Release the version 
5. Run migrations on production databases (not dropping the column)
6. Deploy the version (which now doesn't reference the column)
7. Wait for the next version 
8. Remove the column from the `include_object` function passed to Alembic 
9. Generate a schema change to drop the column
10. Release the version 
11. Run migrations on production databases (Now actually dropping the column)
12. Deploy the version 

The above flow is really easy to get wrong, one has to make seemingly unrelated changes in multiple files and it's quite easy to forget removing the condition ignoring the column from `include_object` function.

**So what are we looking for?**
A dream come true solution would be having a parameter we can pass to the `sa.Column` constructor which would make SQLAlchemy mark the column as _Optional_ or _Under deprecation_. This would:
1. Make it so that SQLAlchemy won't raise an error if the column doesn't exist in the database as it won't try to select/update/insert it. 
2. On the other hand Alembic would still recognize it as being there and so won't generate a schema change for dropping it

This would simplify our flow to: 
1. Remove from code all usage of a model column
2. Specify the column as being deprecated/optional 
3. Release the version 
4. Run migrations on production databases (not dropping the column)
5. Deploy the version (which now doesn't reference the column)
6. Wait for the next version 
7. Delete the column
8. Generate a schema change to drop the column 
9. Release the version 
10. Run migrations on production databases (Now actually dropping the column)
11. Deploy the version 

----

* Is there something like it this that we just haven't been able to find for some reason? 
* Anything else we can do towards getting closer to our desired solution?
* Alternative solutions? 

Thank you in advance for any feedback or advice  
🙏

**Comment by zzzeek** [comment]
so you have an include_object function which has the given column/table inside of, we would assume a list.   and the ideal solution would be, dont have a list, instead, have a directive on Column itself.    and the problem is, that the tablename/colname is now maintained in a place far away from where the model itself is.

You can add things to Column using [Column.info](https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=column+info#sqlalchemy.schema.Column.params.info).  Having SQLAlchemy itself ignore a Column in a model, at least from the ORM side of things, doable by using the [exclude_properties](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=mapper#sqlalchemy.orm.mapper) collection.   The column is still there from a Core perspective but the ORM should not refer to it in statements.   You can consume this dictionary before the mapping is made and turn it into exclude_properties using [mapper_args](https://docs.sqlalchemy.org/en/14/orm/declarative_config.html#mapper-configuration-options-with-declarative) with declared_attr (note however this assumes a 1-1 naming convention with the way the column is declared on the class):

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import Session
Base = declarative_base()



class OmitColsMixin:
    @declared_attr
    def __mapper_args__(cls):
        if cls.__table__ is not None:
            return {"exclude_properties": [
                c.key for c in cls.__table__.c if c.info.get("exclude", False)
            ]}
        else:
            return {}


class A(OmitColsMixin, Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    x = Column(String)
    y = Column(String, info={"exclude": True})

```

you can then consume that same info in your [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) function.

**Reply by ilanKeshet** [reply]
Thank you @zzzeek I'll give that a try 👍
